bin/GenValidation/scripts/Plot_code/histo.py

```diff
@@ -0,0 +1,364 @@
+import os
+import sys
+import optparse
+import ROOT
+import math
+
+from ROOT import TH1F, TFile, TTree, TLorentzVector, TVectorD
+from optparse import OptionParser
+from math import cos, sin, sinh, hypot, pi
+MW_=80.379
+
+class extend_p4:
+  p4_vector=ROOT.TLorentzVector()
+  pdgid=11
+  def __init__(self, ltvector, pid):
+    self.p4_vector=ltvector
+    self.pdgid=pid
+
+def deltaPhi(phi1, phi2):
+  dphi = (phi1 - phi2)
+  while dphi > pi:
+    dphi -= 2 * pi
+  while dphi < -pi:
+    dphi += 2 * pi
+  return dphi
+
+def deltaR(eta1, phi1, eta2, phi2):
+    return hypot(eta1 - eta2, deltaPhi(phi1, phi2))
+
+def wreco(lepp4, metpt, metphi):
+  tmpsol1=1.0
+  tmpsol2=1.0
+  nu4=TLorentzVector()
+  px=metpt*cos(metphi)
+  py=metpt*sin(metphi)
+  pz=0.
+  pxl=lepp4.Px()
+  pyl=lepp4.Py()
+  pzl=lepp4.Pz()
+  El=lepp4.E()
+  a=MW_*MW_ + 2.*pxl*px + 2.*pyl*py
+  A=4.*(El*El - pzl*pzl)
+  B=-4.*a*pzl
+  C=4.*El*El*(px*px+py*py)-a*a
+  tmproot=B*B - 4.*A*C
+
+  if tmproot<0:
+    pz=-1.*B/(2*A)
+  else:
+    tmpsol1 = (-B + math.sqrt(tmproot))/(2.0*A)
+    tmpsol2 = (-B - math.sqrt(tmproot))/(2.0*A)
+    if abs(tmpsol2-pzl)<abs(tmpsol1-pzl):
+      pz=tmpsol2
+    else:
+      pz=tmpsol1
+  if abs(pz)>300:
```

Comment:
Just out of curiosity, is this how W mass is reconstructed usually in our CMS analysis?

Reply:
comparing with the HWW analysis, the methods are the same when no real solution exist, i.e., using the real part of the complex solution. The difference is that in the case of B*B-4AC>0, HWW always uses the smaller solution. But here i use the smaller one only when solution of neutrino pz is quite large, otherwise i use the one closer to the lepton pz.

Reply:
Ah ok thanks :)